tokio-timer/src/timer/entry.rs

```diff
@@ -0,0 +1,556 @@
+        unsafe { (*self.when.get()) }
+    }
+
+    pub fn set_when_internal(&self, when: Option<u64>) {
+        unsafe { (*self.when.get()) = when; }
+    }
+
+    /// Called by `Timer` to load the current value of `state` for processing
+    pub fn load_state(&self) -> Option<u64> {
+        let state = self.state.load(SeqCst);
+
+        if is_elapsed(state) {
+            None
+        } else {
+            Some(state)
+        }
+    }
+
+    pub fn is_elapsed(&self) -> bool {
+        let state = self.state.load(SeqCst);
+        is_elapsed(state)
+    }
+
+    pub fn fire(&self, when: u64) {
+        let mut curr = self.state.load(SeqCst);
+
+        loop {
+            if is_elapsed(curr) || curr > when {
+                return;
+            }
+
+            let next = ELAPSED | curr;
+            let actual = self.state.compare_and_swap(curr, next, SeqCst);
+
+            if curr == actual {
+                break;
+            }
+
+            curr = actual;
+        }
+
+        self.task.notify();
+    }
+
+    pub fn error(&self) {
+        // Only transition to the error state if not currently elapsed
+        let mut curr = self.state.load(SeqCst);
+
+        loop {
+            if is_elapsed(curr) {
+                return;
+            }
+
+            let next = ERROR;
+
+            let actual = self.state.compare_and_swap(curr, next, SeqCst);
+
+            if curr == actual {
+                break;
+            }
+
+            curr = actual;
+        }
+
+        self.task.notify();
+    }
+
+    pub fn cancel(entry: &Arc<Entry>) {
+        let state = entry.state.fetch_or(ELAPSED, SeqCst);
+
+        if is_elapsed(state) {
+            // Nothing more to do
+            return;
+        }
+
+        let inner = match entry.inner.upgrade() {
+            Some(inner) => inner,
+            None => return,
+        };
+
+        let _ = inner.queue(entry);
+    }
+
+    pub fn poll_elapsed(&self) -> Poll<(), Error> {
+        use futures::Async::NotReady;
+
+        let mut curr = self.state.load(SeqCst);
+
+        if is_elapsed(curr) {
+            if curr == ERROR {
+                return Err(Error::shutdown());
+            } else {
+                return Ok(().into());
+            }
+        }
+
+        self.task.register();
+
+        curr = self.state.load(SeqCst).into();
+
+        if is_elapsed(curr) {
+            if curr == ERROR {
+                return Err(Error::shutdown());
+            } else {
+                return Ok(().into());
+            }
+        }
+
+        Ok(NotReady)
+    }
+
+    pub fn reset(entry: &Arc<Entry>, deadline: Instant) {
+        let inner = match entry.inner.upgrade() {
+            Some(inner) => inner,
+            None => return,
+        };
+
+        let when = inner.normalize_deadline(deadline);
+        let elapsed = inner.elapsed();
+
+        let mut curr = entry.state.load(SeqCst);
+        let mut notify;
+
+        loop {
+            if curr == ERROR {
```

Comment:
TIOLI but it seems like this could possibly just be `if curr == ERROR || curr == when { return; }`